feat: use --run flag when executing single test ( run single test instead of all of them )

encero opened the pull request:
Run the go test command with --run flag when executing single test.

Support for neotest `type=test` was removed in https://github.com/nvim-neotest/neotest-go/pull/72

This is unfortunate but i believe in wisdom of the maintainers. Anyway i really need this functionality. Otherwise the plugin is unusable on projects with e2e/integration tests. I can't execute the whole suite each time, that is job for the CI.

---
The solution is untested for all the different nested, map/slice based tests. This works for me right now. This PR is mostly for visibility only. I hope when someone encounter the same roadblock they can use this a inspiration how to solve it. 

Plus multiple people are trying to merge stuff to the same piece of code and i don't want to fight over  it right now. 

Way of avoiding this all together would be providing extension point inside the `build_spec` function for users to change the outcome themselves. Something in line of 

```go
require("neotest-go")({
     test_command_hook = function(command, position) 
         table.insert(command, "--run")
         table.insert(command, position.name)
     end,
      args = { "-count=1", "-timeout=60s" }
    })
```

fredrikaverpil commented:
I just added an issue about this here without realizing there was a PR up for this already: https://github.com/nvim-neotest/neotest-go/issues/83

fredrikaverpil reviewed (approved):
🚀 yes please!

fredrikaverpil commented:
@encero I think the regex looks a bit wrong  for subtests (nested `t.Run()`).

```json
{ "cd", "/Users/fredrik/code/public/go-playground/bugs/neotest-go", "&&", "go", "test", "-v", "-json", "", "-coverprofile=/Users/fredrik/code/public/go-playground/bugs/neotest-go/coverage.out", "--run", '\\^"Level 2"/"Level 3"\\$', "./" }
```

```go
func Test_Level_1(t *testing.T) {
	t.Run("Level 2", func(t *testing.T) {
		t.Run("Level 3", func(t *testing.T) {        // I had my cursor here when I ran "nearest test"
			assert.Equal(t, 1, 1)
		})
	})
}
```

The regex should be `^Test_Level_1/Level_2/Level_3$`, not `\\^"Level 2"/"Level 3"\\$`.

fredrikaverpil commented:
@encero I think the regex looks a bit wrong  for subtests (nested `t.Run()`).

Let's say you have this and run "Level 3":

```go
func Test_Level_1(t *testing.T) {
	t.Run("Level 2", func(t *testing.T) {
		t.Run("Level 3", func(t *testing.T) {  // I had my cursor here when I ran "nearest test"
			assert.Equal(t, 1, 1)
		})
	})
}
```

The neotest-go command then gets generated like this on my end:

```json
{ "cd", "/Users/fredrik/code/public/go-playground/bugs/neotest-go", "&&", "go", "test", "-v", "-json", "", "-coverprofile=/Users/fredrik/code/public/go-playground/bugs/neotest-go/coverage.out", "--run", '\\^"Level 2"/"Level 3"\\$', "./" }
```

Here, the regex does not look right: `\\^"Level 2"/"Level 3"\\$`. The test's _actual_ name is `Test_Level_1/Level_2/Level_3`.

fredrikaverpil commented on lua/neotest-go/init.lua:
```diff
@@ -161,6 +161,12 @@ function adapter.build_spec(args)
+
+  local run_flag = {}
+  if position.type == "test" then
+    run_flag = { "--run", "\\^" .. utils.get_prefix(args.tree, position.name) .. "\\$" }
```
All other arguments use just one dash. Not sure what the difference would be here (if any) but perhaps keep it consistent?

Also, why the backslashes?

```suggestion
    run_flag = { "-run", "^" .. utils.get_prefix(args.tree, position.name) .. "$" }
```

encero replied:
Those slashes are from [here](https://github.com/nvim-neotest/neotest-go/pull/72/files#diff-ba39acc61170d012d9fe7694e11f1e9c7060a687fda264dec455706a38da4a8aL165),  without it works also fine

fredrikaverpil replied:
Not sure what @sergii4 says about them, but I think we don't need those backslashes. And I'd vote for `-run` (one dash) to keep the arguments consistently using one dash.

encero commented:
> Unfortunately, I'm not sure how to figure out dynamically named tests, like this one:

I don't think you can do that without a go "interpreter".

I would expect those kind of test to shop up in the summary window. I they would then a workaround will be to run the whole test func and then select a sub test in the summary window.  But that is not the case, the test doesn't show up in the summary 🤔.

fredrikaverpil commented:
### Dynamically generated sub-tests

> I don't think you can do that without a go "interpreter".

@encero you are right, after some investigation it doesn't seem that we can support dynamically generated sub-tests at all.

> I would expect those kind of test to shop up in the summary window.

The test summary doesn't show the full picture, and you cannot run the "Level 3" test unless you "run nearest" at Level 1 or Level 2.

<img width="1113" alt="image" src="https://github.com/nvim-neotest/neotest-go/assets/994357/471cfb6d-166a-4c3e-bb46-154e2849533c">

So putting that dream to rest, maybe let's focus on landing this PR as it is super useful and I'm also in the same integration test scenario as you, where I cannot have all tests run.

### Regex

> You can maybe avoid doing regex magic by walking the tree?

Maybe. Not sure if it's worth it though, since we get the data we need from the `position`, it just needs to be reformatted, like I did in the regex.

What do you think @sergii4 (see my diff [here](https://github.com/nvim-neotest/neotest-go/pull/81#issuecomment-1962405183))?

encero commented on lua/spec/neotest-go/init_spec.lua:
```diff
@@ -190,6 +190,24 @@ describe("discover_positions", function()
+          },
+          {
+            {
+              id = vim.loop.cwd() .. "/neotest_go/cases_test.go::TestAdd::\"test three\"::test_four",
```
There is something wonky with the treesitter query, those quotes are unexpected to say the least

encero commented:
Found few minutes to spent on this

- got rid of the double dash in run flag
- renamed utils.get_prefix to more accurate name
- more spec tests around this feature
- implemented the "walk tree to get test name" approach

> Maybe. Not sure if it's worth it though, since we get the data we need from the position, it just needs to be reformatted, like I did in the regex.
I personally hate string manipulation, and regexes in general. Sometimes they are unavoidable sure. 

> you are right, after some investigation it doesn't seem that we can figure out what the names are of dynamically runtime-generated sub-tests here.
You could unleash an [ast](https://pkg.go.dev/go/ast) on it, i sure don't have a time for that though.

fredrikaverpil commented:
@encero @sergii4 what are your opinions on the best approach here?

I would really like us to try and move forward here with _any_ kind of implementation that avoids having to maintain a local fork of neotest-go, which is currently what I have to do.

sergii4 commented:
@fredrikaverpil what's the urgency? Why do you have to maintain a local fork ?

sergii4 commented:
@encero could you please fix formatting?

sergii4 commented on lua/spec/neotest-go/init_spec.lua:
```diff
@@ -349,21 +367,48 @@ describe("build_spec", function()
     assert.are.same(expected_command, result.command)
     assert.are.same(path, result.context.file)
   end)
+  -- This test is overwriting plugin global state, keep it at end of the file or face the consequences ¯\_(ツ)_/¯
```
How does it overwrite plugin global state?

sergii4 commented on lua/neotest-go/utils.lua:
```diff
@@ -157,15 +157,22 @@ function utils.get_errors_from_test(test, file_name)
 ---@return string
-function utils.get_prefix(tree, name)
-  local parent_tree = tree:parent()
-  if not parent_tree or parent_tree:data().type == "file" then
```
So we don't need that check anymore, do we?

encero replied:
Which check do you mean?

encero commented:
> @encero could you please fix formatting?

@sergii4 the formatting should be fixed now, at least is green in my fork.
